algebra/src/curves/models/short_weierstrass_jacobian.rs

```diff
@@ -277,6 +277,77 @@ impl<P: Parameters> AffineCurve for GroupAffine<P> {
     fn mul_by_cofactor_inv(&self) -> Self {
         self.mul(P::COFACTOR_INV).into()
     }
+
+    fn apply_endomorphism(&self) -> Self { 
+        let mut self_e = self.clone();
+        self_e.x.mul_assign(P::ENDO_COEFF);
+        self_e
+    }
+
+    fn endo_rep_to_scalar(bits: Vec<bool>) -> Self::ScalarField {
```

Comment:
Mmm not sure of having understood what you mean, the problem is even before all that. Since in the for loop we call bits[i * 2 + 1], if bits are not even, last iteration will cause a panic due to access of a non-existing array index.

Reply:
But the integer division `bits.len()/2` handles this?